### PR TITLE
finality: BIP-32 based randomness derivation

### PR DESCRIPTION
**Pull request by SebastianElvis**
First step towards implementing #591 

This PR introduces BIP-32 based randomness derivation. This includes

- Implementing master secret/public randomness and relevant functions
- Adding master public randomness to `FinalityProvider` DB object
- Adding master public randomness to `MsgCreateFinalityProvider` and modifying the corresponding message handler
- Adding master public randomness to API types

The PR also includes/modifies relevant fuzz/e2e tests.

The subsequent steps include the following and will be done in subsequent PRs

- replacing public randomness committed from `MsgCommitPubRandList` with derived ones
- BTC timestamping master public randomness
-  Allowing finality providers to update master public randomness (TBD)
- Docs and changelogs

### CODE REVIEW

File: proto/babylon/btcstaking/v1/btcstaking.proto [hunk 1]
```diff
@@ -25,14 +25,16 @@ message FinalityProvider {
     bytes btc_pk = 4 [ (gogoproto.customtype) = "github.com/babylonchain/babylon/types.BIP340PubKey" ];
     // pop is the proof of possession of babylon_pk and btc_pk
     ProofOfPossession pop = 5;
+    // master_pub_rand is the master public randomness of the finality provider
+    string master_pub_rand = 6;
```
Comment (gitferry) [inline proto/babylon/btcstaking/v1/btcstaking.proto:29]: here we use `string` type but in `FinalityProviderWithMeta` we use `bytes` type. Maybe we can make it consistent with `string`?

File: x/btcstaking/client/cli/tx.go [hunk 2]
```diff
@@ -108,13 +108,17 @@ func NewCreateFinalityProvicerCmd() *cobra.Command {
...
-				BabylonPk:   &babylonPK,
-				BtcPk:       btcPK,
-				Pop:         pop,
+				Signer:        clientCtx.FromAddress.String(),
+				Description:   &description,
+				Commission:    &rate,
+				BabylonPk:     &babylonPK,
+				BtcPk:         btcPK,
+				Pop:           pop,
+				MasterPubRand: mpr,
```
Comment (gitferry) [inline x/btcstaking/client/cli/tx.go:121]: should we validate the format of the pub rand?

**Review by gitferry — APPROVED** [review]
Great work!

File: proto/babylon/btcstaking/v1/btcstaking.proto (hunk 1 above)
Comment (SebastianElvis) [reply proto/babylon/btcstaking/v1/btcstaking.proto:29]: nice catch! made it as `string` now

File: crypto/eots/randomness.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func NewMasterRandPair(randSource io.Reader) (*MasterSecretRand, *MasterPublicRand, error) {
+	// get random seed
+	var seed [32]byte
+	if _, err := io.ReadFull(randSource, seed[:]); err != nil {
+		return nil, nil, err
+	}
+	// generate new master key pair
+	masterSK, err := hdkeychain.NewMaster(seed[:], &chaincfg.MainNetParams)
```
Comment (KonradStaniec) [inline crypto/eots/randomness.go:40]: Note on `NewMaster` says that we should check for specific err:
```
// NOTE: There is an extremely small chance (< 1 in 2^127) the provided seed
// will derive to an unusable secret key.  The ErrUnusable error will be
// returned if this should occur, so the caller must check for it and generate a
// new seed accordingly.
```

also why hardcoding ` &chaincfg.MainNetParams` ?

File: proto/babylon/btcstaking/v1/query.proto [hunk 4]
```diff
@@ -318,16 +318,18 @@ message FinalityProviderResponse {
   bytes btc_pk = 4 [ (gogoproto.customtype) = "github.com/babylonchain/babylon/types.BIP340PubKey" ];
   // pop is the proof of possession of babylon_pk and btc_pk
   ProofOfPossession pop = 5;
+  // master_pub_rand is the master public randomness of the finality provider
+  string master_pub_rand = 6;
```
Comment (KonradStaniec) [inline proto/babylon/btcstaking/v1/query.proto:322]: I would add comments about the fact that string is base58 encoding

**Review by KonradStaniec — COMMENTED** [review]
just some late comments, nothing critica :) great work 👍

File: crypto/eots/randomness.go (hunk 3 above)
Comment (SebastianElvis) [reply crypto/eots/randomness.go:40]: Good point, will fix in a follow up

> also why hardcoding  &chaincfg.MainNetParams ?

No particular reason tbh. The value is not critical for our use case, so I picked up mainnet. Probably we can leave it as is?

File: proto/babylon/btcstaking/v1/query.proto (hunk 4 above)
Comment (SebastianElvis) [reply proto/babylon/btcstaking/v1/query.proto:322]: Good point, will add in a follow up

File: crypto/eots/randomness.go (hunk 3 above)
Comment (SebastianElvis) [reply crypto/eots/randomness.go:40]: resolved in [`b4b486f` (#595)](https://github.com/babylonchain/babylon/pull/595/commits/b4b486f1e98a124341c461984d3065d52dc83df7)

File: proto/babylon/btcstaking/v1/query.proto (hunk 4 above)
Comment (SebastianElvis) [reply proto/babylon/btcstaking/v1/query.proto:322]: resolved in [`b4b486f` (#595)](https://github.com/babylonchain/babylon/pull/595/commits/b4b486f1e98a124341c461984d3065d52dc83df7)

**Pull request merged**